api/test/service/upload.test.js

```diff
@@ -0,0 +1,311 @@
+    with: space.did(),
+    nb: { root },
+    proofs: [proof]
+    // @ts-expect-error ʅʕ•ᴥ•ʔʃ
+  }).execute(connection)
+
+  // Validate DB
+  const dbItems = await getItemsFromUploadTable(t.context.dynamoClient, space)
+  t.is(dbItems.length, 0)
+})
+
+test('store/list does not fail for empty list', async (t) => {
+  const uploadService = t.context.serviceDid
+  const alice = await Signer.generate()
+  const space = await Signer.generate()
+
+  const connection = await getClientConnection(uploadService, t.context)
+  const proof = await getSpaceDelegationProof(space, alice, space.did())
+  
+  const uploadList = await UploadCapabilities.list.invoke({
+    issuer: alice,
+    audience: uploadService,
+    with: space.did(),
+    proofs: [ proof ]
+    // @ts-expect-error ʅʕ•ᴥ•ʔʃ
+  }).execute(connection)
+
+  t.like(uploadList, { results: [], pageSize: 0 })
+})
+
+test('store/list returns entries previously uploaded by the user', async (t) => {
+  const uploadService = t.context.serviceDid
+  const alice = await Signer.generate()
+  const space = await Signer.generate()
+
+  const connection = await getClientConnection(uploadService, t.context)
+  const proof = await getSpaceDelegationProof(space, alice, space.did())
+
+  // invoke multiple upload/add with proof
+  const cars = [
+    await randomCAR(128),
+    await randomCAR(128)
+  ]
+
+  for (const car of cars) {
+    await UploadCapabilities.add.invoke({
+      issuer: alice,
+      audience: uploadService,
+      with: space.did(),
+      nb: { root: car.roots[0], shards: [car.cid] },
+      proofs: [proof]
+      // @ts-expect-error ʅʕ•ᴥ•ʔʃ
+    }).execute(connection)
+  }
+
+  /** @type {import('../../service/types').ListResponse<import('../../service/types').UploadItemOutput>} */
+  const uploadList = await UploadCapabilities.list.invoke({
+    issuer: alice,
+    audience: uploadService,
+    with: space.did(),
+    proofs: [ proof ]
+    // @ts-expect-error ʅʕ•ᴥ•ʔʃ
+  }).execute(connection)
+
+  t.is(uploadList.pageSize, cars.length)
+
+  // Validate entries have given CARs
+  for (const entry of uploadList.results) {
+    t.truthy(cars.find(car => car.roots[0].toString() === entry.dataCID ))
+  }
+})
+
+/**
+ * 
+ * @param {import('@ucanto/principal/ed25519').EdSigner} service 
+ * @param {any} context 
+ * @returns 
+ */
+async function getClientConnection (service, context) {
+  return UcantoClient.connect({
+    id: service,
+    encoder: CAR,
+    decoder: CBOR,
+    channel: await createTestingUcantoServer(context),
+  })
+}
+
+/**
+ * @param {import("@ucanto/principal/dist/src/ed25519/type.js").EdSigner<"key">} issuer
+ * @param {import("@ucanto/principal/dist/src/ed25519/type.js").EdSigner<"key">} audience
+ * @param {`${string}:${string}`} resource
+ */
+async function getSpaceDelegationProof (issuer, audience, resource) {
+  return await UcantoClient.delegate({
+    issuer,
+    audience,
+    capabilities: [{ can: '*', with: resource }]
+  })
+}
```

Comment:
@olizilla feels way cleaner this way then having to replicate this across the board. what do you think?

Reply:
yep. is good. I would go further and have the function create the `space` signer too, which is what the `createSpace` function of the agent would do.